SeaBIOS support

pietrushnic opened the pull request:
Please let me know how to split those changes to merge most of them. The results are far from production, but I'm working on fixes.

```sh
% ./scripts/ci/qemu-self-test-seabios.sh
==============================================================================
Setup-And-Boot-Seabios :: This suite verifies the correct operation of keyw...
==============================================================================
Enter Boot Menu SeaBIOS :: Test Enter Boot Menu kwd                   | PASS |
------------------------------------------------------------------------------
Enter Boot Menu SeaBIOS And Return Construction :: Test Enter Boot... | PASS |
------------------------------------------------------------------------------
Enter sortbootorder :: Test Enter sortbootorder kwd                   | FAIL |
No match found for 'Save configuration and exit' in 10 seconds. Output:
SeaBIOS (version rel-1.16.0.1-23-g5b5fd6ffPCI: XHCI at 00:04.0 (mmio 0xfcf64000)
XHCI init: regs @ 0xfcf64000, 8 ports, 64 slots, 32 byte contexts
XHCI    protocol USB  2.00, 4 ports (offset 5), def 0
XHCI    protocol USB  3.00, 4 ports (offset 1), def 0
/3ed38000\ Start thread
AHCI controller at 00:1f.2, iobase 0xfcf61000, irq 10
AHCI: cap 0xc0141f05, ports_impl 0x3f
/3ed36000\ Start thread
|3ed36000| AHCI/0: probing
/3ed35000\ Start thread
|3ed35000| AHCI/1: probing
/3ed33000\ Start thread
|3ed33000| AHCI/2: probing
|3ed33000| AHCI/2: link up
|3ed33000| Searching bootorder for: /pci@i0cf8/*@1f,2/drive@2/disk@0
|3ed33000| Searching bios-geometry for: /pci@i0cf8/*@1f,2/drive@2/disk@0
/3ed32000\ Start thread
|3ed32000| AHCI/3: probing
    [ Message content over the limit has been removed. ]
Searching bootorder for: /pci@i0cf8/*@2
Unable to find VPD table
Unable to find VPD table

Press F10 key now for boot menu

Searching bootorder for: HALT
Space available for UMB: dc800-ed000, f69e0-f70e0
Returned 16773120 bytes of ZoneHigh
e820 map has 7 items:
  0: 0000000000000000 - 000000000009fc00 = 1 RAM
  1: 000000000009fc00 - 00000000000a0000 = 2 RESERVED
  2: 00000000000f0000 - 0000000000100000 = 2 RESERVED
  3: 0000000000100000 - 000000003fd45000 = 1 RAM
  4: 000000003fd45000 - 0000000040000000 = 2 RESERVED
  5: 00000000b0000000 - 00000000c0000000 = 2 RESERVED
  6: 00000000fed40000 - 00000000fed45000 = 2 RESERVED
enter handle_19:

Booting from ROM..Booting from ca00:0385
.
iPXE starting execution...ok
iPXE initialising devices...unimplemented handle_1aXX:258:
    [ Message content over the limit has been removed. ]
Trying next device: 2
Booting from DVD/CD..AHCI/2: ... finished, status 0x41, ERROR 0x20
Device reports MEDIUM NOT PRESENT - 2 tries left
AHCI/2: ... finished, status 0x41, ERROR 0x20
Device reports MEDIUM NOT PRESENT - 1 tries left
AHCI/2: ... finished, status 0x41, ERROR 0x20
Device reports MEDIUM NOT PRESENT - 0 tries left
AHCI/2: ... finished, status 0x41, ERROR 0x20
.
Boot failed: Could not read from CDROM (code 0003enter handle_18:
  NULL
)
Trying next device: enter handle_18:
  NULL
3
Trying next device: enter handle_18:
  NULL
4
Trying next device: 5
No bootable device.  Retrying in 60 seconds.
------------------------------------------------------------------------------
Setup-And-Boot-Seabios :: This suite verifies the correct operatio... | FAIL |
5 tests, 3 passed, 2 failed
==============================================================================
```

The two most important fixes support for flash in sortbootorder and figuring out how to run iPXE in QEMU, so QEMU would not use OptionROM,  but SeaBIOS will load builtin one.

pietrushnic committed:
add lib/bios/seabios.robot, remove resource menus.robot

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs/include/default.robot: disable OPTIONS_LIB, which depends on menus.robot and conflict seabios

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs/qemu-selftests.robot: change configuration to one used by seabios

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-run.sh: give ability to provide QEMU_FW_FILE through env var

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-run.sh: enabled kvm only if it supported

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-self-test-seabios.sh: initial commit

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
self-tests/setup-and-boot-seabios.robot: initial commit

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts: add support for Dasharo (coreboot+SeaBIOS) for QEMU Q35

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic commented:
@macpijan I wonder how far I should go with that, now results are as follows:

```sh
% ./scripts/ci/qemu-self-test-seabios.sh
==============================================================================
Setup-And-Boot-Seabios :: This suite verifies the correct operation of keyw...
==============================================================================
Enter Boot Menu SeaBIOS :: Test Enter Boot Menu kwd                   | PASS |
------------------------------------------------------------------------------
Enter Boot Menu SeaBIOS And Return Construction :: Test Enter Boot... | PASS |
------------------------------------------------------------------------------
Enter sortbootorder :: Test Enter sortbootorder kwd                   | PASS |
------------------------------------------------------------------------------
Enter TPM Configuration :: Test Enter TPM Configuration kwd           | PASS |
------------------------------------------------------------------------------
Enter iPXE :: Test Enter iPXE kwd                                     | FAIL |
No match found for 'autoboot' in 10 seconds. Output:
------------------------------------------------------------------------------
Setup-And-Boot-Seabios :: This suite verifies the correct operatio... | FAIL |
5 tests, 4 passed, 1 failed
==============================================================================
Output:  /home/user/src/3mdeb/dasharo/open-source-firmware-validation/logs/2024.11.12_14.29.14/setup-and-boot-seabios/output.xml
Log:     /home/user/src/3mdeb/dasharo/open-source-firmware-validation/logs/2024.11.12_14.29.14/setup-and-boot-seabios/log.html
Report:  /home/user/src/3mdeb/dasharo/open-source-firmware-validation/logs/2024.11.12_14.29.14/setup-and-boot-seabios/report.html
zsh: exit 1     ./scripts/ci/qemu-self-test-seabios.sh
```

Of course, I can satisfy all the checks despite the self-test change, and if my approach is correct, I would probably need your approval.

Regarding the SeaBIOS self-test, my concern is that I'm entering the land of testing `sortbootorder`; some tests exist in the legacy framework, and I could start porting those, but then the question arises: what is the difference between the keyword that should be tested and test itself. I think we should validate iPXE and network booting comprehensively, but to enter iPXE you need to enable network, which means you have to teach OSFV how to walk into `sortbootorder` (already done ) and change some option (probably done in legacy test suite).

pietrushnic committed:
lib/bios/seabios.robot: add more keywords and its tests

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic commented:
Another set of changes:

```sh
% ./scripts/ci/qemu-self-test-seabios.sh
==============================================================================
Setup-And-Boot-Seabios :: This suite verifies the correct operation of keyw...
==============================================================================
Enter Boot Menu SeaBIOS :: Test Enter Boot Menu kwd                   | PASS |
------------------------------------------------------------------------------
Enter Boot Menu SeaBIOS And Return Construction :: Test Enter Boot... | PASS |
------------------------------------------------------------------------------
Enter sortbootorder :: Test Enter sortbootorder kwd                   | PASS |
------------------------------------------------------------------------------
Get sortbootorder Menu Construction :: Get sortbootorder Menu Cons... | PASS |
------------------------------------------------------------------------------
Get Option State :: Test Get Option State kwd                         | PASS |
------------------------------------------------------------------------------
Enter Menu From Snapshot and Return sortbootorder Consruction :: T... | PASS |
------------------------------------------------------------------------------
Enter TPM Configuration :: Test Enter TPM Configuration kwd           | PASS |
------------------------------------------------------------------------------
Enter iPXE :: Test Enter iPXE kwd                                     | FAIL |
No match found for 'autoboot' in 10 seconds. Output:
------------------------------------------------------------------------------
Setup-And-Boot-Seabios :: This suite verifies the correct operatio... | FAIL |
8 tests, 7 passed, 1 failed
==============================================================================
Output:  /home/user/src/3mdeb/dasharo/open-source-firmware-validation/logs/2024.11.14_00.54.30/setup-and-boot-seabios/output.xml
Log:     /home/user/src/3mdeb/dasharo/open-source-firmware-validation/logs/2024.11.14_00.54.30/setup-and-boot-seabios/log.html
Report:  /home/user/src/3mdeb/dasharo/open-source-firmware-validation/logs/2024.11.14_00.54.30/setup-and-boot-seabios/report.html
zsh: exit 1     ./scripts/ci/qemu-self-test-seabios.sh
```

pietrushnic commented:
```sh
% ./scripts/ci/qemu-self-test-seabios.sh
==============================================================================
Setup-And-Boot-Seabios :: This suite verifies the correct operation of keyw...
==============================================================================
Enter Boot Menu SeaBIOS :: Test Enter Boot Menu kwd                   | PASS |
------------------------------------------------------------------------------
Enter Boot Menu SeaBIOS And Return Construction :: Test Enter Boot... | PASS |
------------------------------------------------------------------------------
Enter sortbootorder :: Test Enter sortbootorder kwd                   | PASS |
------------------------------------------------------------------------------
Get sortbootorder Menu Construction :: Get sortbootorder Menu Cons... | PASS |
------------------------------------------------------------------------------
Get Option State :: Test Get Option State kwd                         | PASS |
------------------------------------------------------------------------------
Enter Menu From Snapshot and Return sortbootorder Consruction :: T... | PASS |
------------------------------------------------------------------------------
Enter TPM Configuration :: Test Enter TPM Configuration kwd           | PASS |
------------------------------------------------------------------------------
Enable Network Boot :: Test Enable Network/PXE boot                   | PASS |
------------------------------------------------------------------------------
Enter iPXE :: Test Enter iPXE kwd                                     | PASS |
------------------------------------------------------------------------------
Setup-And-Boot-Seabios :: This suite verifies the correct operatio... | PASS |
9 tests, 9 passed, 0 failed
==============================================================================
```

pietrushnic committed:
add support for BIOS_LIB uefi or seabios

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
.github/workflows/qemu-self-test-seabios.yml: initial commit

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

macpijan commented on lib/bios/seabios.robot:
```diff
@@ -0,0 +1,305 @@
+    [Documentation]    Enter given sortbootorder Menu option and return construction
+    [Arguments]    ${menu}    ${option}
+    ${key}=    Extract Sortbootorder Menu Key    ${menu}    ${option}
+    Write Bare Into Terminal    ${key}
```
Link to the same potential issue with PiKVM where `Write Bare Into Terminal ` is used. I can see in UEFI we have always this IF/ELSE. It can be resolved in a better way, but not necessarily in this PR.

macpijan commented on lib/bios/seabios.robot:
```diff
@@ -0,0 +1,305 @@
+    [Documentation]    This keyword saves introduced changes
+    Write Bare Into Terminal    s
+
+Get IPXE Boot Menu Construction
```
The same as UEFI

pietrushnic replied:
@macpijan there are too many leftovers I have clean that up: a9ac47cb55f7

macpijan replied:
Ok. Please ping me once you believe you have a complete list of keywords you need, or at least remove those that are not needed at the time, so we do not waste time discussing something to be dropped anyway.

macpijan replied:
@pietrushnic I saw a new MR, do you consider this one finalized per your current needs?

pietrushnic replied:
@macpijan Yes, I need to figure out how to extend it without starting to enable new tests. If you think something is missing here, considering there is minimal support and new PRs coming, please let me know, and I will try to improve.

Feedback on whether the approach is clean enough would also be helpful.

macpijan commented on scripts/ci/qemu-self-test.sh:
```diff
@@ -2,13 +2,13 @@
-  "robot -L TRACE -v config:qemu-selftests -v rte_ip:127.0.0.1 -d ./logs/$(date +%Y.%m.%d_%H.%M.%S)/list-options -v snipeit:no self-tests/list-options.robot"
-  "robot -L TRACE -v config:qemu-selftests -v rte_ip:127.0.0.1 -d ./logs/$(date +%Y.%m.%d_%H.%M.%S)/secure-boot -v snipeit:no self-tests/secure-boot.robot"
-  "robot -L TRACE -v config:qemu-selftests -v rte_ip:127.0.0.1 -d ./logs/$(date +%Y.%m.%d_%H.%M.%S)/self-tests -v snipeit:no self-tests/terminal.robot"
+  "robot -X -L TRACE -v BIOS_LIB:uefi -v config:qemu-selftests-uefi -v rte_ip:127.0.0.1 -d ./logs/$(date +%Y.%m.%d_%H.%M.%S)/setup-and-boot-menus -v snipeit:no self-tests/setup-and-boot-menus.robot"
```
BIOS_LIB can be added to platform config instead. No need to add it from command line, if we have two separate configs anyway.

I would suggest to set BIOS_LIB to uefi in `platform-configs/include/default.robot` and only override with seabios one in seabios qemu config.

pietrushnic replied:
e383a2555926

macpijan committed:
platform-configs/qemu: swap release and selftests configs

It appers to me that the selftest config should have more menus
enabled, than the "release" config, supporting only "reasonable"
menus.

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
platform-configs/qemu-selftests-uefi.robot: use common part from qemu,robot

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan commented on platform-configs/qemu-selftests-seabios.robot:
```diff
@@ -12,7 +12,8 @@ ${INITIAL_DUT_CONNECTION_METHOD}=       Telnet
 ${DUT_CONNECTION_METHOD}=               ${INITIAL_DUT_CONNECTION_METHOD}
 ${RTE_S2_N_PORT}=                       1234
 ${FLASH_SIZE}=                          ${8*1024*1024}
-${BOOT_MENU_KEY}=                       ${ESC}
```
Apply something like: https://github.com/Dasharo/open-source-firmware-validation/pull/574/commits/413581a80c38d79ff5a822216dd155ce6282a336

pietrushnic replied:
e122d83ac04f

macpijan commented on scripts/ci/qemu-run.sh:
```diff
@@ -57,7 +61,8 @@ This is the QEMU wrapper script for the Dasharo Open Source Firmware Validation.
 Example usage:
-    ./$(basename $0) vnc firmware
+    ./$(basename $0) vnc uefi
+    ./$(basename $0) nographic seabio
```
```suggestion
    ./$(basename $0) nographic seabios
```

macpijan commented on scripts/ci/qemu-self-test-seabios.sh:
```diff
@@ -0,0 +1,21 @@
+
+# Define an array of commands
+commands=(
+  "robot -X -L TRACE -v BIOS_LIB:seabios -v config:qemu-selftests-seabios -v rte_ip:127.0.0.1 -d ./logs/$(date +%Y.%m.%d_%H.%M.%S)/setup-and-boot-seabios -v snipeit:no self-tests/setup-and-boot-seabios.robot"
```
BIOS_LIB can be set in platform config

pietrushnic replied:
fixed in https://github.com/Dasharo/open-source-firmware-validation/commit/e383a2555926adf42b2cf5838c65d640e76cc204

macpijan reviewed (changes requested):
I think more code can be reused, rather than copied in two places.

First thought would be to have:
- lib/bios/common.robot - some common keywords, that can be reused across differnt bios versions
- lib/bios/seabios.robot - seabios keywords - either entirely new ones, or calling ones from common. robot, maybe with some extra arguments
- lib/bios/edk2.robot (?) - for dasharo uefi menus based on edk2

pietrushnic commented on .github/workflows/qemu-self-test-seabios.yml:
```diff
@@ -0,0 +1,53 @@
+name: Keywords self-tests for Dasharo (coreboot+SeaBIOS) on QEMU Q35
```
We should stick to GH's actions. The question is if this would be reliable. Sooner or later, we will exhaust the pool of free actions and storage. I don't see ourselves switching to paid GH, so I would switch to our Woodpecker CI instead.

But I get your point about reusability. My gh-actions-fu was not good enough to be aware of that possibility. There is no space for such work for now, but we should track it as a potential improvement. Based on your comment, I will create an issue for the following PC Engines release.

pietrushnic committed:
keywords.robot: do not import lib/bios/menus.robot twice

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic commented on keywords.robot:
```diff
@@ -551,6 +550,19 @@ Prepare Test Suite
+    [Documentation]    Import BIOS libraries based on config and command
+    ...    line variables
+    IF    '${BIOS_LIB}' == 'uefi'
+        Import Resource    ${CURDIR}/lib/bios/menus.robot
```
a1f41211be7edadb726c370851607dec387e6375

pietrushnic commented:
> I think more code can be reused, rather than copied in two places.
> 
> First thought would be to have:
> 
>     * lib/bios/common.robot - some common keywords, that can be reused across differnt bios versions
> 
>     * lib/bios/seabios.robot - seabios keywords - either entirely new ones, or calling ones from common. robot, maybe with some extra arguments
> 
>     * lib/bios/edk2.robot (?) - for dasharo uefi menus based on edk2

I like those ideas and will work on implementing them. However, I'm worried that PR will grow indefinitely and never be merged. I have time to push now, but I may return to it in 2-3 months. Nobody will pick improvements, and we will face various issues. I would be glad to divide it into smaller PRs, which consist of changes that can already be accepted and will not destroy anything. Still, I will slowly move the development of SeaBIOS forward, delivering something each time.

pietrushnic committed:
platform-configs/qemu-selftests-uefi.robot: fix resource order

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic commented on lib/bios/seabios.robot:
```diff
@@ -0,0 +1,305 @@
+    ${menu}=    Get Setup Menu Construction
+    RETURN    ${menu}
+
+Get RTC Clock Submenu Construction
```
No, it is not the only one; there are at least two more, but those are "hidden":
- https://github.com/pcengines/sortbootorder?tab=readme-ov-file#hidden-security-registers-menu
- https://github.com/pcengines/sortbootorder?tab=readme-ov-file#hidden-flash-lockdown-menu

pietrushnic replied:
removed in b43573901928 as unused

pietrushnic committed:
platform-configs,scripts: make BIOS_LIB=uefi default and override only in qemu seabios config

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs/qemu-selftests-seabios.robot: import common qemu.robot resource

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic commented on lib/bios/seabios.robot:
```diff
@@ -0,0 +1,305 @@
+*** Settings ***
+Documentation       Collection of keywords related to SeaBIOS.
+
+# Here is comparison of terms with lib/bios/menus.robot
```
I can do this, but it would require a force push, which will close all your comments. Maybe this should be the last step. Do you have a better idea of how to address this comment?

macpijan replied:
We can also ignore it, I have already diffed manually

pietrushnic replied:
Shall we close this?

macpijan commented:
Let's see what can be done in this iteration. Move the rest to the next one.

pietrushnic committed:
docs,scripts: keep firmware option for qemu-run.sh as backward compatibility

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic commented on scripts/ci/qemu-run.sh:
```diff
@@ -176,8 +183,17 @@ esac
 
 case "${ACTION}" in
-  firmware)
+  uefi)
```
Please let me know if this is enough 842a3440646d
I left `firmware` for backward compatibility.

I don't see qemu-run.sh in the context of `firmware` mentioned in dasharo docs. It is mentioned in the context of `os` and `os_install`.

macpijan commented on lib/bios/seabios.robot:
```diff
@@ -0,0 +1,305 @@
+    #    Title line, such as:    Dasharo System Features
+    #    BOTTOM:
+    #    Help line, such as:    F9=Reset to Defaults    Esc=Exit
+    ${submenu}=    Get Menu Construction    ${checkpoint}    ${lines_top}    ${lines_bot}
```
This keyword is not defined aywhere? Since `Get RTC Clock Submenu Construction` uses unused keyword, I assume it has not been tested. 

Which keyword from this file do you actually need to execute the tests, and which have been tested @pietrushnic ?

macpijan commented on lib/bios/seabios.robot:
```diff
@@ -0,0 +1,305 @@
+
+    RETURN    ${value}[0]
+
+Set Option State And Return Construction
```
This is also not used in self-tests, or in any tests so far. Has it been tested? I am confused which keywords are really needed for seabios operation. Please confirm this.

pietrushnic committed:
rename lib/bios/menus.robot -> lib/bios/edk2.robot

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
use platform-defined FW_STRING instead of {SEABIOS,TIANOCORE}_STRING

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/common.robot: initial support for common Enter Boot Menu kwd

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic commented:
> * lib/bios/edk2.robot (?) - for dasharo uefi menus based on edk2

@macpijan https://github.com/Dasharo/open-source-firmware-validation/pull/574/commits/fd2d51fdf3de858d32e7193d4f9671574a61258b

pietrushnic commented:
> * lib/bios/common.robot - some common keywords, that can be reused across differnt bios versions

This commit starts standard lib: c3fc40a27728
I moved `Enter Boot Menu {Tianocore,SeaBIOS}` to `Enter Boot Menu` and parameterized with `FW_STRING` instead of `{SEABIOS,TIANOCORE}_STRING`.

pietrushnic committed:
add unified Enter Boot Menu and Return Construction

This commit replace:
- Enter Boot Menu Tianocore And Return Construction
- Enter Boot Menu SeaBIOS And Return Construction
with one keyword Enter Boot Menu and Return Construction, which is
provided by lib/bios/common.robot

It also move Parse Menu Snapshot Into Construction to common lib.

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/seabios.robot: remove unused Get IPXE Boot Menu Construction

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/seabios.robot: remove unused Get RTC Clock Submenu Construction

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
unify Enter Setup Menu kwd for uefi and seabios

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios: user Get Menu Construction instead of SeaBIOS dedicated wrapper

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/seabios.robot: remove unused keywords

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/edk2.robot: remove Get Menu Consutruction use common implmentation

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/seabios.robot: align API with edk2 lib

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios: move Get Index Of Matching Option In Menu to common

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

filipleple committed:
lib/bios: move Boot System Or From Connected Disk to common lib

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
platform-configs/optiplex-7010-seabios.robot: add

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
lib/bios: make Boot System Or From Connected Disk more generic

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

pietrushnic commented:
@filipleple, how was the "Boot System ..."  keyword tested? I guess you didn't automatically install OS because of this https://github.com/Dasharo/open-source-firmware-validation/pull/274 ?

macpijan commented:
> @filipleple, how was the "Boot System ..." keyword tested? I guess you didn't automatically install OS because of this https://github.com/Dasharo/open-source-firmware-validation/pull/274 ?

We install operating systems (typically Windows and Ubuntu) when setting up the platforms using configs from: https://github.com/Dasharo/preseeds

pietrushnic commented:
@macpijan, so "setting up the platform" is not automated? I'm not sure what that means. Is this Assembly in Dasharo Lab or preparation before running regression/release tests?

I want to leverage `os_install` option from qemu-run.sh and perform automatic tests on top of Dasharo (coreboot+SeaBIOS) for QEMU. Are you suggesting I use `create_image.sh` and then provide it as `INSTALLER_PATH`? 

I prefer an automatic installation mechanism that can be kicked as a test, but the difference between those approaches may be negligible. In that case, the remaining arguments for automatic installation in tests would be:
- excluding humans from the process to limit potential errors,
- limit additional steps,
- decrease maintenance overhead (we should update to 24.04.1) and keep adjustments in one place (OSFV) instead of OSFV and pressed repo

We used to have automatic OS installation, and I used that in legacy PC Engines tests. I would like to know how we plan to address this topic better to boost release cycle time. Shouldn't we have tests where we install over netboot.xyz with our preseed?

macpijan commented:
> so "setting up the platform" is not automated? I'm not sure what that means. Is this Assembly in Dasharo Lab or preparation before running regression/release tests?

We typically install Windows -> Ubuntu using automated installers from preseeds repo when setting up platform in the lab.It must be in this order to install them both on one disk.

> I want to leverage os_install option from qemu-run.sh and perform automatic tests on top of Dasharo (coreboot+SeaBIOS) for QEMU. Are you suggesting I use create_image.sh and then provide it as INSTALLER_PATH?

You could, or you could test the proposal from Daniil for Ubuntu Server installer. For PC Engines you are probably not that interested in the Ubuntu Desktop installer. Also, you do not have the limitations caused by Windows here.

filipleple commented:
Yes, I've used a platform with pre-installed OS'es, as we usually do with regression/testing tasks 

Here's the full log for dmidecode for reference: https://cloud.3mdeb.com/index.php/s/mALHsbN7DgWeSpx
